src/bindings/python/flux/uri/resolvers/lsf.py

```diff
@@ -0,0 +1,102 @@
+
+It can be used like:
+  `flux uri --remote lsf:<JOBID>`.
+
+"""
+
+
+def lsf_find_compute_node(jobid) -> str:
+    """Figure out where the job is being run using JSON output from `bjobs -l`."""
+
+    """
+    To get the name of the system (lassen or sierra or nid) without the node number,
+    pull the hostname and strip off anything that's not alphabetic.
+    """
+    host_system = subprocess.run(["hostname"], capture_output=True).stdout.decode(
+        "utf-8"
+    )
+    host_system = re.sub(r"[^a-zA-Z]", "", host_system)
+
+    """
+    A way more efficient way to do this could be revealed if we knew what API calls
+    LSF was making. There is API documentation in the LSF docs, but LSF is closed source.
+    """
+    sp = subprocess.run(["bjobs", "-l", str(jobid)], capture_output=True)
+    job_output = re.sub(r"\s", "", sp.stdout.decode("utf-8"))  ## removes whitespace
+    nodes = re.findall(
+        rf"[0-9]+\*{host_system}[0-9]+", job_output
+    )  ## returns list of possible nodes
+    """
+    Apply a logic to distinguish between batch nodes and launch nodes (that is, batch should be)
+    using more cores than launch regardless of center policies on this. Also, this logic should work at centers
+    that don't use launch nodes at all (unknown if OLCF does).
+    """
+    hostname, host_cores = "", 0
+    for node in nodes:
+        node = node.split("*")
+        if int(node[0]) > host_cores:
+            hostname = node[1]
+    if hostname != "":
+        return hostname
+    raise ValueError("Unable to resolve the hostname of the compute job.")
+
+
+def lsf_job_pid(jobid, hostname) -> str:
+    """Resolve the pids of an LSF job using ssh to the compute node"""
+    sp = subprocess.run(
+        ["ssh", hostname, "ps", "-Hu", "`whoami`", "|", "grep", "-m1", "flux-broker"],
+        capture_output=True,
+    )
+    if sp.returncode != 0:
+        raise ValueError(f"Unable to find a flux session on host {hostname}")
+    return sp.stdout.decode("utf-8").split()[0]
+
+
+def lsf_get_uri(hostname, pid) -> str:
+    """
+    Resolve the URI of a Flux instance by running
+
+      `ssh <HOSTNAME> flux uri --remote pid:<PID>`
+
+    and converting the result from bytecode to a string.
+    """
+    sp = subprocess.run(
+        ["ssh", hostname, "flux", "uri", "--remote", f"pid:{pid}"], capture_output=True
+    )
+    return sp.stdout.decode("utf-8")
+
+
+class URIResolver(URIResolverPlugin):
+    """A URIResolver that can fetch a FLUX_URI from an LSF job id"""
+
+    def describe(self):
+        return "Get URI for a Flux instance launched under IBM Spectrum LSF"
+
+    def resolve(self, uri):
+        jobid = str(PurePath(uri.path).parts[0])
+
+        try:
+            hostname = lsf_find_compute_node(jobid)
+            return lsf_get_uri(hostname, lsf_job_pid(jobid, hostname))
+        except (ValueError, OSError):
+            pass
+        raise ValueError(f"LSF Job {jobid} doesn't seem to have a FLUX_URI")
```

Comment:
I'm not sure I understand the point of suppressing an exception and then raising a different one. You lose the traceback that way, and the actual error could be completely unrelated to jobids/FLUX_URI. I would suggest either not trying to catch the exception at all or something like:
```python
try:
    ...
except (ValueError, OSError) as exc:
    raise ValueError("LSF job doesn't have FLUX_URI") from exc
```
https://stackoverflow.com/questions/24752395/python-raise-from-usage

Reply:
This could be my fault if the practice was copied from the `slurm` jobid resolver. Apologies if so!

Reply:
I'd say it's definitely not good practice to put error handling wherever possible. Often the best thing to do is to _not_ catch exceptions, especially in lower-level (i.e. simpler) functions. Off the top of my head I would say you should only catch an exception if

- You can work around the error and still give the desired functionality. For instance, you're pinging some unreliable service, and it fails. The proper thing to do might be to catch the error and try again rather than letting the error propagate.
- The error message won't make any sense to the user of the function and so you want to change the error message and type. But in this case I think it's good to `raise ... from` because then you can see both the original error and the final error with the more intelligible message.

I think you would develop a feel for it as you write more Python, and you see cases where error messages are not helpful at all, and some that are quite helpful.

Reply:
Since this is a user tool I would think about what kinds of error messages would be helpful to users, who might not know much about programming, or about Python. I can think of a few:

- jobid is not valid
- there isn't LSF on the cluster (i.e. it's Slurm or something else. probably very rare)
- there isn't a Flux instance running under the job

If you can detect one of those conditions, I think it would be nice to throw a simple error and don't try to catch it anywhere, just let it propagate and terminate Python. Otherwise, if it's something weird and nitty-gritty going wrong, like the YAML returned by CSM doesn't have the right entry, I would just let the 'normal' error happen because the true source of the issue is going to be  opaque to users no matter what, and at least if you let the normal error happen (as opposed to catching it and rethrowing  a generic "something went wrong!" error) you'll get a message that's useful for developers.

Reply:
Having said so much I'm pretty sure I'm bound to have said something wrong so if someone else can correct me (@grondo ) please do.